mathphys/imgproc.py

```diff
@@ -0,0 +1,1282 @@
+        #   2.7 ms ± 23.2 µs per loop
+        #   (mean ± std. dev. of 7 runs, 100 loops each)
+        self._fitx = None
+        self._fity = None
+        self._angle = 0
+        self._fitgauss = fitgauss or FitGaussian
+        super().__init__(*args, **kwargs)
+        self._update_image_fit(roix=roix, roiy=roiy)
+
+    @property
+    def fity(self):
+        """."""
+        return self._fity
+
+    @property
+    def fitx(self):
+        """."""
+        return self._fitx
+
+    @property
+    def roiy(self):
+        """."""
+        return self.fity.roi
+
+    @roiy.setter
+    def roiy(self, value):
+        """."""
+        self._update_image_fit(roix=self.fitx.roi, roiy=value)
+
+    @property
+    def roix(self):
+        """."""
+        return self.fitx.roi
+
+    @roix.setter
+    def roix(self, value):
+        """."""
+        self._update_image_fit(roix=value, roiy=self.fity.roi)
+
+    @property
+    def roi(self):
+        """."""
+        return self.fitx.roi, self.fity.roi
+
+    @roi.setter
+    def roi(self, value):
+        """."""
+        self._update_image_fit(*value)
+
+    @property
+    def angle(self):
+        """Image tilt angles [deg]."""
+        return self._angle
+
+    def update_roi_with_fwhm(self, fwhmx_factor=2, fwhmy_factor=2):
+        """."""
+        self.fitx.update_roi_with_fwhm(fwhm_factor=fwhmx_factor)
+        self.fity.update_roi_with_fwhm(fwhm_factor=fwhmy_factor)
+
+    def calc_angle_with_roi(self, sigma_factor, nrpts):
+        """."""
+        # generate posx grid
+        cx, sx = self.fitx.roi_mean, self.fitx.roi_sigma
+
+        posx = _np.linspace(cx - sigma_factor*sx, cx + sigma_factor*sx, nrpts)
+        posx = list(set([int(val) for val in posx]))
+        posx = _np.sort(posx)
+
+        images1droi = [Image1D_ROI(data=self.data[:, val]) for val in posx]
+        posy = [image.roi_center for image in images1droi]
+
+        pfit = _np.polynomial.polynomial.polyfit(posx, posy, 1)
+        angle = - _np.arctan(pfit[1]) # sign due to vertical dir pixel increase
+        angle *= 180 / _np.pi
+
+        return angle
+
+    def calc_mode_sigmas(self):
+        """."""
+        # method:
+        #
+        # [x, y] = R(-angle) [u1, u2]
+        #
+        # R(-angle) = [[C, S], [-S, C]]
+        #
+        # sigmax² = C² sigma1² + S² sigma2²
+        # sigmay² = S² sigma1² + C² sigma2²
+        #
+        # TODO: not working...
+        sigmax = self.fitx.roi_sigma
+        sigmay = self.fity.roi_sigma
+        sigma_max = max(sigmax, sigmay)
+        func, funs = _np.cos(-self.angle), _np.sin(-self.angle)
+
+        det = func**2 - funs**2
+        if abs(det) < 1e-6:
+            return _np.nan, _np.nan
+
+        sigma1sqr = func**2 * sigmax**2 - funs**2 * sigmay**2
+        sigma2sqr = -funs**2 * sigmax**2 + func**2 * sigmay**2
+
+        if sigma1sqr > 0:
+            sigma1 = min(_np.sqrt(sigma1sqr), sigma_max)
+        else:
+            sigma1 = _np.nan
```

Comment:
I don't understand why you do this.

Reply:
this is to avoid numerical singularity when image fitted angle is too close to 45 deg. 
when in this situation I want to set principal-axis sigmas to `np.nan` in order to flag unreliable calculation.  

On other cases, where image angle is not close to 45 deg, I take the `min` in order to bound numerical errors on sigma to obey the fact that normal mode sigmas cannot be bigger than the max(sigmax, sigmay) of the tilted image.

Reply:
Ok, now I understand the check for the determinant, but I'm not sure if this covers all cases, because it's sigmax and sigmay are too close the angle is undetermined, cc isn't it? I'm this case you still can calculate the normal modes sigmas?

On the point related to the min, I don't know this property... How/why is it true?

Reply:
in the normal axes the difference between the diagonal second moment matrix elements will always be greater than the corresponding value for an arbitrary rotated set of axes. the rotation increases the lower sigma and reduces the higher sigma. juts think of projected bivariate gaussian along these for axes.